pkg/autoscaler/aggregation/bucketing.go

```diff
@@ -87,6 +108,49 @@ func roundToNDigits(n int, f float64) float64 {
 	return math.Floor(f*p) / p
 }
 
+const (
+	// minExponent is the minimal decay multiplier for the weighted average computations.
+	minExponent = 0.2
+	// The sum of weights for weighted average should be at least this much.
+	weightPrecision = 0.9999
+	precision       = 6
+)
+
+// WeightedAverage returns the exponential weighted average
+// TODO(vagababov): optimize for O(1) computation, if possible.
+func (t *TimedFloat64Buckets) WeightedAverage(now time.Time) float64 {
+	if t.IsEmpty(now) {
+		return 0
+	}
+
+	now = now.Truncate(t.granularity)
+	t.bucketsMutex.RLock()
+	defer t.bucketsMutex.RUnlock()
+	totalB := len(t.buckets)
+	numB := len(t.buckets)
+
+	multiplier := t.decayMultiplier
+	// We start with 0es. But we know that we have _some_ data because
+	// IsEmpty returned false.
```

Comment:
is it OK to rely on this even though t.IsEmpty(now) happened outside the lock?

Reply:
Probably OK, but I'll make sure we hold the lock for hte duration.